src/test/util/setup_common.cpp

```diff
@@ -120,18 +119,39 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", "Custom data directory (default: <random_string>)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    if (!m_node.args->IsArgSet("-testdatadir")) {
+        // By default, the data directory has a random name
+        const std::string testdatadir{g_insecure_rand_ctx_temp_path.rand256().ToString().c_str()};
+        m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / testdatadir;
+    } else {
+        // Custom data directory
+        const std::string testdatadir{m_node.args->GetArg("-testdatadir").value()};
+        m_path_root = fs::path(testdatadir.c_str()) / "test_temp";
+        fs::remove_all(m_path_root);
+
+        // Print the test directory name if custom, but only once per test run
+        static bool g_printed_dir{false};
+        if (!g_printed_dir) {
+            g_printed_dir = true;
+            std::cout << "Test directory (will not be deleted): " << m_path_root << std::endl;
+        }
```

Comment:
Is this because you are running all tests within a boost fixture test suite or is it happening when you run a single test case? 
E.g. fixture test suite: `./test_bitcoin --run_test=db_tests` vs single test case `./test_bitcoin --run_test= db_tests/getwalletenv_file`.

Reply:
The `g_printed_dir` guard is only needed because the user may be running more than one test case (like your first example). I don't love this global variable but it seems weird to print this out on every individual test case, and I think it's okay for a test program. I convinced myself that no race condition is possible, but we could use `std::call_once()` here but I didn't think it was worth it.

Reply:
> The `g_printed_dir` guard is only needed because the user may be running more than one test case 

This would break the PR goal in the current implementation. Only the last executed test datadir would be kept, the rest will be deleted.
I think we should include the test name in the path and create a datadir directory for each executed tests. Could pull most of what I did before: 53eb4980d561491798d0c8b0ee383c83ba193432.